File: src/main/java/scrabble/phrases/words/Noun.java
package scrabble.phrases.words;

import java.util.Objects;

/**
 * Romanian noun with gender and articulation support.
 *
 * @param word the noun in base form
 * @param gender the grammatical gender
 * @param syllables the number of syllables
 * @param rhyme the rhyme suffix (last 3 characters)
 */
public record Noun(
    String word,
    NounGender gender,
    int syllables,
    String rhyme
) implements Word {

    /**
     * Creates a new Noun with computed syllables and rhyme.
     *
     * @param word the noun in base form
     * @param gender the grammatical gender
     */
    public Noun(String word, NounGender gender) {
        this(
            word,
            gender,
            WordUtils.computeSyllableNumber(word),
            WordUtils.computeRhyme(word)
        );
    }

    /**
     * Compact constructor for validation.
     */
    public Noun {
        Objects.requireNonNull(word, "word cannot be null");
        Objects.requireNonNull(gender, "gender cannot be null");
    }

    /**
     * Gets the articulated form of the noun.
     *
     * @return the articulated form
     */
    public String articulated() {
        return switch (gender) {
            case M, N -> articulateMasculine();
            case F -> articulateFeminine();
        };
    }

    private String articulateMasculine() {
        if (word.endsWith("u")) {
            return word + "l";
        }
        return word + "ul";
    }

    private String articulateFeminine() {
        if (word.length() > 1 && (word.endsWith("ă") || word.endsWith("ie"))) {
            return word.substring(0, word.length() - 1) + "a";
        }
        if (word.endsWith("a")) {
            return word + "ua";
        }
        return word + "a";
    }

    // Legacy method names for compatibility
    public String getWord() {
        return word;
    }

    public NounGender getGender() {
        return gender;
    }

    public int getSyllables() {
        return syllables;
    }

    public String getRhyme() {
        return rhyme;
    }

    public String getArticulated() {
        return articulated();
    }
}

File: src/main/kotlin/scrabble/phrases/words/Noun.kt
package scrabble.phrases.words

data class Noun(
    override val word: String,
    val gender: NounGender,
    override val syllables: Int = WordUtils.computeSyllableNumber(word),
    override val rhyme: String = WordUtils.computeRhyme(word),
    val articulated: String = computeArticulated(word, gender)
) : Word {
    init {
        require(!word.isBlank()) { "Noun word must not be blank or whitespace-only" }
        val validPattern = Regex("^[a-zA-Z\\u0103\\u0102\\u00E2\\u00CE\\u00EE\\u0218\\u0219\\u021A\\u021B]+$")
        require(validPattern.matches(word)) { "Noun word contains invalid characters: '$word' (allowed: a-z, A-Z, ă/Ă, â/Î/î, Ș/ș, Ţ/ţ)" }
    }

    companion object {
        fun computeArticulated(word: String, gender: NounGender): String = when (gender) {
            NounGender.M, NounGender.N -> articulateMasculine(word)
            NounGender.F -> articulateFeminine(word)
        }

        private fun articulateMasculine(word: String): String = when {
            word.endsWith("u") -> word + "l"
            word.endsWith("e") -> word + "le"
            word.endsWith("ă") -> word + "l"
            else -> word + "ul"
        }

        private fun articulateFeminine(word: String): String = when {
            word.length > 1 && (word.endsWith("ă") || word.endsWith("ie")) ->
                word.dropLast(1) + "a"
            word.endsWith("a") -> word + "ua"
            else -> word + "a"
        }
    }
}
